### ISSUE TITLE
Bluebird promises returned from Waterline missing lots of useful functions

### ISSUE THREAD

**Issue by blah238** [balderdashy/waterline#810]
I'm using Sails 0.10.5 with Waterline 0.10.16, Bluebird 2.3.11 and sails-disk 0.10.7.

If I do something like this in the Sails console:

```
User.findOne({ id: 1}).populate('pets').then(function (user) { return user.pets; }).each(console.log)
```

I get `TypeError: Object [object Promise] has no method 'each'`.

Similarly, `map`, `reduce`, `filter`, `any` are all missing. There are probably others.

Am I doing something wrong or is this a known issue?

**Comment by tjwebb** [comment]
I think you're misunderstanding the bluebird promise API. Once you say `.then()`, then promise is done. `.map` and others can be used **instead** of `.then`, however.

**Comment by blah238** [comment]
Respectfully, I must disagree -- based on their example [here](https://github.com/petkaantonov/bluebird/blob/master/API.md#eachfunction-iterator---promise) it should work:

```
function loadStory() {
  return getJSON('story.json')
    .then(function(story) {
      addHtmlToPage(story.heading);
      return story.chapterURLs.map(getJSON);
    })
    .each(function(chapter) { addHtmlToPage(chapter.html); })
    .then(function() { addTextToPage("All done"); })
    .catch(function(err) { addTextToPage("Argh, broken: " + err.message); })
    .then(function() { document.querySelector('.spinner').style.display = 'none'; });
}
```

Also, these both work fine:

```
var Promise = require('bluebird');
Promise.resolve([1, 2, 3]).then().each(console.log);
```

```
var Promise = require('bluebird');
Promise.each(User.findOne({ id: 1}).populate('pets').then(function (user) { return user.pets; }), console.log)
```

So it would appear that the aforementioned functions are just not there, not that the usage is incorrect. I haven't looked into the Waterline source yet to confirm this, but is it the case that only a subset of the Bluebird API is exposed through it?

**Comment by tjwebb** [comment]
>       return story.chapterURLs.map(getJSON);

They are returning more promises. That's why it works. The promise is fulfilled when you cease returning new promises and return a value, which is what you are doing. Hope that helps.

**Comment by blah238** [comment]
Are you sure? The API documentation explains that it also works on values:

> Iterate over an array, or a promise of an array, which contains promises (or a mix of promises and values)

I really don't think it matters whether their `getJSON` function returns a promise or a value in this case. My other examples confirm this.

I think it's more a matter of not being able to use the chaining syntax due to the missing functions than the validity of the usage at this point. I can work with it, but it's not as nice as it could be.

**Comment by tjwebb** [comment]
Yes I'm sure. `.map` does work on values, if those values are inside a promised array. You are not returning a Promise of an array, you are returning just a regular array.

**Comment by tjwebb** [comment]
https://github.com/tjwebb/sails-permissions/blob/master/config/fixtures/role.js#L16-L24

Does spread work for you?

**Comment by blah238** [comment]
I think `spread` would only work if the number of pets was fixed and known in advance, no? Unless I was to work with the `arguments` array, which I don't think is their intention.

Also the only reason I'm just returning `user.pets` is that their [`get` utility function](https://github.com/petkaantonov/bluebird/blob/master/API.md#getstring-propertynameint-index---promise) doesn't exist either.

Ideally I'd be able to do:

```
User.findOneById(1).populate('pets')
  .then()
  .get('pets')
  .each(doSomethingAsyncWithPet)
  .then(sendResponse);
```

**Comment by tjwebb** [comment]
I'll look into this at some point. We're just using vanilla bluebird promises, so I'm not sure how all these huge problems would exist and no one else has reported them. Or maybe they have and I've missed them.

**Comment by blah238** [comment]
Okay, thanks! I don't think many people are using Bluebird promises with Waterline yet because other than the 0.11.x RCs there isn't an npm package of Sails with a recent enough version of Waterline. I had to install it manually. Side issue though.

**Comment by tjwebb** [comment]
Hmm, if you have to install manually, that's a separate issue.

**Comment by blah238** [comment]
Here are the keys of the Promise object returned by Waterline. These look like Q functions still. For example "promiseDispatch" is not in Bluebird but it is in Q.

```
all
allResolved
allSettled
catch
del
delay
delete
denodeify
dispatch
done
fail
fapply
fbind
fcall
fin
finally
get
inspect
invoke
isFulfilled
isPending
isRejected
join
keys
mapply
mcall
nbind
nfapply
nfbind
nfcall
ninvoke
nmapply
nmcall
nodeify
npost
nsend
passByCopy
post
progress
promiseDispatch
race
send
set
spread
then
thenReject
thenResolve
timeout
toString
valueOf
```

**Comment by tjwebb** [comment]
I didn't implement the Promise switchover; it's possible that the
implementer only added .catch() somehow. I'll look into it. Thanks.

**Comment by blah238** [comment]
My bad TJ, I think I figured it out. My global Sails install was still running an old version of Waterline, so even though my local Waterline has Bluebird, when I was testing with `sails console`, it was still using Q promises. I'm going to play around with it a bit more but it looks like `each`, `map`, `reduce`, `get`, etc. are all working.

**Comment by tjwebb** [comment]
ha. I'm so much less confused now. Thanks for clearing that up.
